docker/machine.go

```diff
@@ -0,0 +1,246 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package docker
+
+import (
+	"encoding/base64"
+	"fmt"
+	"strings"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"sigs.k8s.io/cluster-api-provider-docker/cloudinit"
+	"sigs.k8s.io/kind/pkg/cluster/constants"
+	"sigs.k8s.io/kind/pkg/cluster/nodes"
+	"sigs.k8s.io/kind/pkg/exec"
+)
+
+const (
+	defaultImage = "kindest/node:v1.15.0"
+)
+
+// Machine implement a service for managing the docker containers hosting a kubernetes nodes.
+type Machine struct {
+	logr.Logger
+	cluster   string
+	machine   string
+	container *nodes.Node
+}
+
+// NewMachine returns a new Machine service for the given Cluster/DockerCluster pair.
+func NewMachine(cluster, machine string, logger logr.Logger) (*Machine, error) {
+	if cluster == "" {
+		return nil, errors.New("cluster is required when creating a docker.Machine")
+	}
+	if machine == "" {
+		return nil, errors.New("machine is required when creating a docker.Machine")
+	}
+	if logger == nil {
+		return nil, errors.New("logger is required when creating a docker.Machine")
+	}
+
+	container, err := getContainer(
+		withLabel(clusterLabel(cluster)),
+		withName(machineContainerName(cluster, machine)),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Machine{
+		cluster:   cluster,
+		machine:   machine,
+		Logger:    logger,
+		container: container,
+	}, nil
+}
+
+// ContainerName return the name of the container for this machine
+func (m *Machine) ContainerName() string {
+	return machineContainerName(m.cluster, m.machine)
+}
+
+// ProviderID return the provider identifier for this machine
+func (m *Machine) ProviderID() string {
+	return fmt.Sprintf("docker:////%s", m.ContainerName())
+}
+
+// NodeName return the name of the Kubernetes node corresponding to this machine
+func (m *Machine) NodeName() string {
```

Comment:
can we replace calls to NodeName with direct calls to `m.ContainerName()` instead? I'm not sure this function is doing anything unless NodeName is clearer in some places than ContainerName?